### PR TITLE
Add launchMode parameter to ToggleCommandPalette action

### PR DESCRIPTION
**Pull request by Don-Vito**
## PR Checklist
* [x] Closes https://github.com/microsoft/terminal/issues/8322
* [x] CLA signed. 
* [x] Tests added/passed
* [x] Documentation updated - https://github.com/MicrosoftDocs/terminal/pull/202
* [x] Schema updated.
* [x] I've discussed this with core contributors already. 

## Detailed Description of the Pull Request / Additional comments
Added an optional launchMode parameter to "commandPalette" command. 
The values of the launchMode are either "action" (default) or "command line".

## Validation Steps Performed
* Manual tests

### CODE REVIEW

File: src/cascadia/LocalTests_SettingsModel/KeyBindingsTests.cpp-9d33445a
Comment (carlos-zamora) [inline src/cascadia/LocalTests_SettingsModel/KeyBindingsTests.cpp-9d33445a:1]: You probably wanna remove this file 😉

File: src/cascadia/TerminalApp/CommandPalette.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.cpp:789]:
```suggestion
        const auto mode = (launchMode == CommandPaletteLaunchMode::CommandLine) ?
```
nit

File: src/cascadia/TerminalApp/CommandPalette.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalApp/CommandPalette.cpp:791]: Could/Should we consolidate these into one `CommandPaletteMode` enum in ActionArgs.idl? That way we don't have to do this translation? Or is that a weird thing to do since the mode that the command palette is in isn't really a setting. (this one is more of a thought exercise than a "requested change")

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp
Comment (carlos-zamora) [inline src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:297]: I could be wrong, but I think we actually want this to stay the same? Looking at `MoveFocus` and that needs an arg, but is still defined here.

**Review by carlos-zamora — COMMENTED** [review]
Great work! @zadjii-msft will probably want to take a look at this. But this is a weird week because of Thanksgiving, so it may be a bit before some other team members get to it.

I'll approve after some discussion on the `CommandPaletteMode` thing below.

File: src/cascadia/TerminalApp/CommandPalette.cpp
Comment (DHowett) [reply src/cascadia/TerminalApp/CommandPalette.cpp:791]: Thought exercise already completed in https://github.com/microsoft/terminal/issues/8322#issuecomment-729976112 and https://github.com/microsoft/terminal/issues/8322#issuecomment-729979167

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:297]: I think there is a lack of consistency, since `MoveFocus` is defined but `SetColorScheme` is not. I am not sure this is important as the code path is not reachable, but probably, if there is a bug in GenerateName I would prefer the command not to appear at all rather than go with the default name.

File: src/cascadia/TerminalApp/CommandPalette.cpp
Comment (Don-Vito) [reply src/cascadia/TerminalApp/CommandPalette.cpp:791]: As Dustin mentioned, we didn't want to expose the entire enum, so I thought that the introduction of the new one was the simplest way to achieve this.

File: src/cascadia/LocalTests_SettingsModel/KeyBindingsTests.cpp-9d33445a
Comment (Don-Vito) [reply src/cascadia/LocalTests_SettingsModel/KeyBindingsTests.cpp-9d33445a:1]: Ouch.. sorry 🤦

**Comment by Don-Vito** [comment]
> 
> 
> Great work! @zadjii-msft will probably want to take a look at this. But this is a weird week because of Thanksgiving, so it may be a bit before some other team members get to it.
> 
> I'll approve after some discussion on the `CommandPaletteMode` thing below.

Thanks for the review! Will get to the fixes soon.
Mike already gave me a heads up about his vacation - so no pressure, this will be just yet another one waiting for him :smile:
And probably it's not just this week being weird -  with all the holidays and PTOs we might be back to normal just in mid-January :smile:
Thanks again and happy Thanksgiving!

File: src/cascadia/TerminalApp/CommandPalette.h
Comment (DHowett) [inline src/cascadia/TerminalApp/CommandPalette.h:36]: nit: no need for the `&` here since it's just as expensive to copy a reference as it is to copy a primitive :smile:

File: src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp
Comment (DHowett) [reply src/cascadia/TerminalSettingsModel/ActionAndArgs.cpp:297]: Agreed.

**Review by DHowett — APPROVED** [review]
Love this. Thanks so much.

Only nits, but since you have to rev the PR for code formatting reasons, I'll still sign off :smile:

**Review by DHowett — APPROVED** [review]

**Comment by DHowett** [comment]
(obligated to wait for docs pr, but this time we're only at one approval)

**Comment by Don-Vito** [comment]
> 
> 
> (obligated to wait for docs pr, but this time we're only at one approval)

Added documentation (though something very basic, but I believe it is better this way).

**Comment by Don-Vito** [comment]
@zadjii-msft - if you are here, please take a look :blush: (I know there are conflicts to resolve)

**Review by zadjii-msft — APPROVED** [review]
Yep, this looks good to me. Thanks!

**Review by DHowett — APPROVED** [review]

**Comment by Don-Vito** [comment]
@DHowett - it seems that automerge fails

**Comment by DHowett** [comment]
Oop lemme kick that build. Sorry!

**Comment by Don-Vito** [comment]
Gods of CI please be kind with this one.. don't let it fail randomly..

**Comment by zadjii-msft** [comment]
@ msftbot re-run the CI till it passes

I kid, we really should do something about these CI failures

**Comment by Don-Vito** [comment]
> 
> 
> @ msftbot re-run the CI till it passes
> 
> I kid, we really should do something about these CI failures

@zadjii-msft - it worked!  I guess it is a combination of my prayers and the undocumented msftbot APIs 😆

**Comment by Jspooner09** [comment]
Have computer virus


Jimmy R. 

CONFIDENTIALITY: The information contained in this transmission may contain privileged and confidential information. It is intended only for the use of the person(s) named above. If you are not the intended recipient, you are hereby notified that any review, dissemination, distribution or duplication of this communication, and the information contained in it, is strictly prohibited. If you are not the intended recipient, please contact the sender and immediately destroy all copies of the original message.



On Thursday, December 3, 2020, 11:54 AM, Don-Vito <notifications@github.com> wrote:





@ msftbot re-run the CI till it passes

I kid, we really should do something about these CI failures


@zadjii-msft - it worked! I guess it is a combination of my prayers and the undocumented msftbot APIs 😆

**Pull request merged**
